### PR TITLE
feat(tooltip): Tooltip use MdPanel API

### LINKED ISSUE #9563: mdPanel: Rewrite mdTooltip to use the panel.
Tracking issue for using the mdPanel in tooltip. Any related issues that appear during the migration should refer to this issue as a primary use case.

### PR DESCRIPTION
**Pull request by bradrich**
Tooltip has been refactored to use the MdPanel API to position, animate open and closed, and show the tooltip as a result of user interaction.

Fixes #9563

Ping @ErinCoughlan @crisbeto

### CODE REVIEW

File: src/components/tooltip/tooltip.js [hunk 1]
```diff
@@ -3,391 +3,426 @@
...
+        var debouncedOnResize = $$rAF.throttle(updatePosition);
+
+        // Initialize the theming of the tooltip.
+        $mdTheming(element);
+
+        // Set defaults
+        setDefaults();
+
+        console.debug(element.html().trim());
+        console.debug(scope.mdDirection);
```
Comment (DevVersion) [inline src/components/tooltip/tooltip.js]: Two debug statements

File: src/components/tooltip/tooltip.js [hunk 2]
```diff
@@ -3,391 +3,426 @@
...
-            showTimeout = null;
-
-            if (!scope.visibleWatcher) {
-              onVisibleChanged(scope.visible);
+        function setVisible(value) {
+          // Break if passed value is already in queue or there is no queue and
+          // passed value is current in the controller.
+          if (setVisible.queued && setVisible.value === !!value ||
+              !setVisible.queued && scope.mdVisible === !!value) {
+                return;
```
Comment (DevVersion) [inline src/components/tooltip/tooltip.js]: Looks like an invalid indention here

```js
if (setVisible.queued && setVisible.value === !!value || 
    !setVisible.queued && scope.mdVisible === !!value) {
  return;
}
```

File: src/components/tooltip/tooltip.js [hunk 3]
```diff
@@ -406,9 +441,9 @@ function MdTooltipRegistry() {
   };
 
   /**
-   * Global event handler that dispatches the registered
-   * handlers in the service.
-   * @param  {Event} event Event object passed in by the browser.
+   * Global event handler that dispatches the registered handlers in the
+   * service.
+   * @param {Event} event Event object passed in by the browser
```
Comment (DevVersion) [inline src/components/tooltip/tooltip.js]: As per Erin's feedback last time, I learned that JSDoc like this should be marked as not-nullable

```
{!Event}
```

Same for below?

File: src/components/tooltip/tooltip.scss [hunk 4]
```diff
@@ -6,66 +6,69 @@ $tooltip-top-margin-lg: rem(1.4) !default;
 $tooltip-top-margin-sm: rem(2.4) !default;
 $tooltip-lr-padding-lg: rem(0.8) !default;
 $tooltip-lr-padding-sm: rem(1.6) !default;
-$tooltip-max-width: rem(3.20) !default;
+$tooltip-max-width: rem(3.2) !default;
```
Comment (DevVersion) [inline src/components/tooltip/tooltip.scss]: While you are at it, why using root em here? EM is actually for font / typography

File: src/components/tooltip/tooltip.spec.js [hunk 5]
```diff
@@ -1,20 +1,38 @@
...
-    $$mdTooltipRegistry = _$$mdTooltipRegistry_;
-  }));
+  // var injectLocals = function($injector) {
+  //   $compile = $injector.get('$compile');
+  //   $rootScope = $injector.get('$rootScope');
+  //   $material = $injector.get('$material');
+  //   $timeout = $injector.get('$timeout');
+  //   $mdPanel = $injector.get('$mdPanel');
+  //   $$mdTooltipRegistry = $injector.get('$$mdTooltipRegistry');
+  // };
```
Comment (DevVersion) [inline src/components/tooltip/tooltip.spec.js]: This comment should be removed

File: src/components/tooltip/tooltip.spec.js [hunk 6]
```diff
@@ -1,20 +1,38 @@
...
+
+  beforeEach(function() {
+    module(
+      'material.components.tooltip',
+      'material.components.button',
+      'material.components.panel'
+    );
+    // inject(injectLocals);
```
Comment (DevVersion) [inline src/components/tooltip/tooltip.spec.js]: Same for this one.

File: src/components/tooltip/tooltip.spec.js [hunk 7]
```diff
@@ -1,20 +1,38 @@
...
+  });
+
+  beforeEach(inject(function(_$compile_, _$rootScope_, _$material_, _$timeout_,
+      _$mdPanel_, _$$mdTooltipRegistry_){
+        $compile = _$compile_;
+        $rootScope = _$rootScope_;
+        $material = _$material_;
+        $timeout = _$timeout_;
+        $mdPanel = _$mdPanel_;
+        $$mdTooltipRegistry = _$$mdTooltipRegistry_;
```
Comment (DevVersion) [inline src/components/tooltip/tooltip.spec.js]: I propose using the `$injector` here

```
$compile = $injector.get('$compile');
```

No need for the ugly underscore prefix / suffix

File: src/components/tooltip/tooltip.spec.js [hunk 8]
```diff
@@ -1,20 +1,38 @@
...
+      }));
+
   afterEach(function() {
-    // Make sure to remove/cleanup after each test
     var scope = element && element.scope();
-    scope && scope.$destroy();
+    scope && scope.$destroy;
```
Comment (DevVersion) [inline src/components/tooltip/tooltip.spec.js]: This seems to not call the destroy method?

**Comment by ThomasBurleson** [comment]
Closing until ready for final review. Please submit a new PR at that point.

**Pull request closed without merging**